src/mozanalysis/frequentist_stats/sample_size.py

```diff
@@ -391,7 +398,86 @@ def _get_sample_size_col(col):
         pop_percent = 100.0 * (sample_size / len(df))
         results[col] = {
             "sample_size_per_branch": sample_size,
-            "number_of_clients_targeted": len(df),
             "population_percent_per_branch": pop_percent,
+            "number_of_clients_targeted": len(df),
         }
     return results
+
+
+def variable_enrollment_length_sample_size_calc(
+    bq_context: BigQueryContext,
+    start_date: Union[str, datetime],
+    max_enrollment_days: int,
+    analysis_length: int,
+    metric_list: List[Metric],
+    target_list: List[Segment],
+    variable_window_length: int = 7,
+    experiment_name: str = "",
+    app_id: Optional[str] = "",
+    to_pandas: bool = True,
+    **sizing_kwargs,
+):
+    if variable_window_length > max_enrollment_days:
+        raise ValueError(
+            "Enrollment window length is larger than the max enrollment length."
+        )
+
+    ht = HistoricalTarget(
+        start_date=start_date,
+        analysis_length=analysis_length,
+        num_dates_enrollment=max_enrollment_days,
+        experiment_name=experiment_name,
+        app_id=app_id,
+    )
+
+    df = ht.get_single_window_data(
+        bq_context=bq_context, metric_list=metric_list, target_list=target_list
+    )
+
+    def _get_time_intervals(start_date: Union[str, datetime], interval: int):
+
+        if isinstance(start_date, str):
+            start_date = datetime.strptime(start_date, "%Y-%m-%d")
+
+        date = start_date + timedelta(interval - 1)
+        date_map = {0: date.date()}
+        i = 1
+        while date < start_date + timedelta(days=(max_enrollment_days - 1)):
+            date = date + timedelta(interval)
+            date_map[i] = date.date()
+            i += 1
+
+        date_map[i - 1] = df["enrollment_date"].max()
+        return date_map
+
+    date_intervals = _get_time_intervals(start_date, variable_window_length)
+
+    def _for_interval_sample_size_calculation(i):
+
+        df_interval = df.loc[df["enrollment_date"] < date_intervals[i]]
+        res = z_or_t_ind_sample_size_calc(
+            df=df_interval, metrics_list=metric_list, test="t", **sizing_kwargs
+        )
+        final_res = {}
+        for key in res.keys():
+            final_res[key] = {
+                "enrollment_end_date": date_intervals[i],
+                **res[key],
+            }
+
+        return final_res
+
+    results_dict = {}
+    for m in metric_list:
+        results_dict[m.name] = []
+
+    for i in range(len(date_intervals)):
+        res = _for_interval_sample_size_calculation(i)
```

Comment:
If I were writing this, I'd change the function signature from an `int` (which represents an index into an external map of date intervals), to the date interval itself. Like:

```python
for date_interval in date_intervals:
    res = _for_interval_sample_size_calculation(date_interval)
```

Because then above you don't need need to index into the date_intervals dict at all, which limits the sharing of state between this function and the outer scope.

Reply:
Actually, upon further reading, it doesn't seem like we really need to key for the date_intervals dict, so what if instead we went with an array of end dates. So `_get_time_intervals` could be renamed as `_get_end_dates` and return a `List[date]`, this loop could be rewritten as:
```python
end_dates = _get_end_dates(start_date, variable_window_length)

for end_date in end_dates:
    res = _for_interval_sample_size_calculation(end_date)

```

Reply:
Ah, agreed. The dictionary with end dates is an artifact of a previous way I was doing this, where the intervals do matter. I'll change it to a list and disregard the keys.